opengever/maintenance/scripts/find_fields_with_non_persisted_default_values.py

```diff
@@ -0,0 +1,200 @@
+"""
+Script to find field values that haven't been persisted on objects.
+
+    bin/instance run find_fields_with_non_persisted_default_values.py
+
+This script logs a detailed CSV report and a summary to var/log/, and displays
+some progress info and stats on STDERR/STDOUT.
+"""
+
+from App.config import getConfiguration
+from collections import Counter
+from datetime import datetime
+from opengever.base.default_values import get_persisted_value_for_field
+from opengever.maintenance.debughelpers import setup_app
+from opengever.maintenance.debughelpers import setup_option_parser
+from opengever.maintenance.debughelpers import setup_plone
+from plone import api
+from plone.dexterity.utils import iterSchemataForType
+from zope.component import getUtility
+from zope.intid.interfaces import IIntIds
+from zope.schema import getFieldsInOrder
+import logging
+import os
+import sys
+import transaction
+
+
+root_logger = logging.root
+
+
+class NonPersistedValueFinder(object):
+
+    CSV_HEADER = "intid;portal_type;path;created;missing_fields"
+    SCHEMA_CACHE = {}
+    FIELD_CACHE = {}
+
+    def __init__(self):
+        self.catalog = api.portal.get_tool('portal_catalog')
+        self.intids = getUtility(IIntIds)
+
+        self.stats = Counter()
+        self.stats['by_field'] = Counter()
+
+        ts = datetime.today().strftime('%Y-%m-%d_%H-%M-%S')
+        self.csv_log_path = self.get_logfile_path(
+            'find-nonpersistent-values-%s.csv' % ts)
+        self.summary_log_path = self.get_logfile_path(
+            'find-nonpersistent-values-summary-%s.log' % ts)
+
+    def run(self):
+        sys.stderr.write("Checking for non-persisted values...\n\n")
+
+        all_brains = self.catalog.unrestrictedSearchResults()
+        total = len(all_brains)
+
+        with open(self.csv_log_path, 'w') as self.csv_log:
+            with open(self.summary_log_path, 'w') as self.summary_log:
+                self.csv_log.write(self.CSV_HEADER + '\n')
+
+                for i, brain in enumerate(all_brains):
+                    obj = brain.getObject()
+                    missing_fields = self.check_for_missing_fields(obj)
+                    self.update_stats(missing_fields)
+
+                    if missing_fields:
+                        self.write_csv_row(obj, missing_fields)
+
+                    if i % 100 == 0:
+                        sys.stderr.write("Progress: %s of %s objects\n" % (i, total))
```

Comment:
i would prefer a `logger` here, there are examples in other scripts how to set one up that is working. directly writing to stdout/stderr is a bit of an antipattern IMO.

Reply:
Right. I copied the existing script and made minimal adjustments to the logic to only get non persisted fields that actually have a default value. But you're right I'll change that